sdk/core/AzureCore/Source/Pipeline/Transport/HTTPRequest.swift

```diff
@@ -48,7 +48,10 @@ public class HTTPRequest: DataStringConvertible {
 
     public init(method: HTTPMethod, url: String, headers: HTTPHeaders, files: [String]? = nil, data: Data? = nil) {
         self.httpMethod = method
-        self.url = url
+        guard let encodedUrl = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
```

Comment:
I worry about doing the encoding here for a couple reasons. First, if I supply a URL to the initializer and then read the `url` property, they will be different. Also, if I set a new URL via that property, it won't be encoded.

Can we delay any sort of normalization like this until just before the request is actually dispatched (presumably in URLSessionTransport)?

In my experience, other workarounds like adding a setter on the `url` property that encodes the URL are brittle, because it's too easy to accidentally double-encode something.

This should also allow library authors to avoid having to fiddle with the encoding like you were doing below in StorageStreamDownloader.

Reply:
Makes sense. I'll take a look. This was needed because otherwise URLs with space in them failed.

Reply:
Fixed. The StorageStreamDownloader did really need URL encoding--it just needed to get the last part of a blob name that might have a "virtual folder" portion of the name (ex:  blob named "virtualdir/file.txt")